File: src/middlewares/captcha.ts
import { Request, Response, NextFunction } from '@tinyhttp/app'
import { HTTPError } from '../errors'
import fetch from 'node-fetch'
import config from '../config'
import { GenerateGuard } from '@itchatapp/controllers'

export const captcha = () => async (req: Request, _res: Response, next: NextFunction): Promise<void> => {
  const key = req.body.captcha_key

  if (!key) {
    throw new HTTPError('FAILED_CAPTCHA')
  }

  const payload = {
    secret: config.captcha.token,
    response: key,
    sitekey: config.captcha.key
  }

  const response = await fetch('https://hcaptcha.com/siteverify', {
    method: 'POST',
    body: JSON.stringify(payload),
  }).then((res) => res.json()) as { success: boolean }

  if (!response || !response.success) {
    throw new HTTPError('FAILED_CAPTCHA')
  }

  next()
}

export const Captcha = () => GenerateGuard(captcha())

File: src/middlewares/captcha.rs
use crate::config::*;
use axum::{
    http::{Request, StatusCode},
    middleware::Next,
    response::Response,
};
use serde::Deserialize;

#[derive(Deserialize)]
struct CaptchaResponse {
    success: bool,
}

pub async fn handle<B>(req: Request<B>, next: Next<B>) -> Result<Response, StatusCode> {
    let key = req.headers().get("X-Captcha-Key");

    if key.is_none() {
        return Err(StatusCode::BAD_REQUEST);
    }

    let client = reqwest::Client::new();
    let body = serde_json::json!({
        "response": key.unwrap().to_str().unwrap(),
        "secret": *CAPTCHA_TOKEN,
        "sitekey": *CAPTCHA_KEY
    });

    let res = client
        .post("https://hcaptcha.com/siteverify")
        .body(body.to_string())
        .send()
        .await;

    if let Ok(res) = res {
        if let Ok(captcha) = res.json::<CaptchaResponse>().await {
            if captcha.success {
                return Ok(next.run(req).await);
            }
        }
    }

    Err(StatusCode::BAD_REQUEST)
}
